File: 2019/go/day_02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	dat, _ := ioutil.ReadFile("2019/inputs/day_02.txt")
	ops := parseOpcodes(dat)

	// Reset the program to the "1202 program alarm" state
	ops[1] = 12
	ops[2] = 2

	star1(ops)
}

func star1(ops []int) {
	i := 0
	exited := false
	for exited != true {
		switch ops[i] {
		case 1:
			arg1 := ops[ops[i+1]]
			arg2 := ops[ops[i+2]]
			outi := ops[i+3]
			ops[outi] = arg1 + arg2
		case 2:
			arg1 := ops[ops[i+1]]
			arg2 := ops[ops[i+2]]
			outi := ops[i+3]
			ops[outi] = arg1 * arg2
		case 99:
			exited = true
		}
		i += 4
	}
	fmt.Printf("%v\n", ops)
}

func parseOpcodes(input []byte) []int {
	ops := []int{}
	for _, str := range strings.Split(string(input), ",") {
		num, _ := strconv.Atoi(strings.Replace(str, "\n", "", 1))
		ops = append(ops, num)
	}
	return ops
}

File: 2019/nim/day_02.nim
from strformat import fmt
from strutils import parseInt, replace, split
from sequtils import insert, map

var input = readFile("2019/inputs/day_02.txt").replace("\n", "").split(",").map(parse_int)

proc run(noun: int, verb: int): int =
  var ops = input.deep_copy
  ops[1] = noun
  ops[2] = verb

  var pc = 0
  var op = ops[pc]
  while op != 99:
    let arg1 = ops[ops[pc + 1]]
    let arg2 = ops[ops[pc + 2]]
    let pos  = ops[pc + 3]
    case op:
      of 1:
        ops[pos] = arg1 + arg2
      of 2:
        ops[pos] = arg1 * arg2
      else:
        echo fmt"Unknown operator '{op}' at index '{pc}'"
        raise
    pc += 4
    op = ops[pc]
  return ops[0]

proc star_1(input: seq[int]): int =
  run(noun = 12, verb = 2)

proc star_2(input: seq[int]): int =
  for noun in 0..99:
    for verb in 0..99:
      if run(noun, verb) == 19690720:
        return 100 * noun + verb

echo star_1(input)
echo star_2(input)
